include/fiction/algorithms/simulation/sidb/critical_temperature.hpp

```diff
@@ -0,0 +1,378 @@
+//
+// Created by Jan Drewniok on 06.02.23.
+//
+
+#ifndef FICTION_CRITICAL_TEMPERATURE_HPP
+#define FICTION_CRITICAL_TEMPERATURE_HPP
+
+#include "calculating_energy_and_state_type.hpp"
+#include "fiction/algorithms/simulation/sidb/energy_distribution.hpp"
+#include "fiction/algorithms/simulation/sidb/exhaustive_ground_state_simulation.hpp"
+#include "fiction/algorithms/simulation/sidb/occupation_probability_excited_states.hpp"
+#include "fiction/technology/cell_technologies.hpp"
+#include "fiction/technology/charge_distribution_surface.hpp"
+#include "fiction/technology/sidb_charge_state.hpp"
+#include "fiction/traits.hpp"
+#include "fiction/types.hpp"
+#include "fiction/utils/hash.hpp"
+#include "fiction/utils/math_utils.hpp"
+#include "fiction/utils/truth_table_utils.hpp"
+
+#include <fmt/format.h>
+#include <kitty/bit_operations.hpp>
+#include <kitty/dynamic_truth_table.hpp>
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <cstdint>
+#include <limits>
+#include <map>
+#include <ostream>
+#include <utility>
+#include <vector>
+
+namespace fiction
+{
+
+/**
+ * This struct stores the parameters for the `critical_temperature` algorithm.
+ */
+struct critical_temperature_params
+{
+    /**
+     * Gate based simulation is conducted.
+     */
+    bool gate_based_simulation = true;
+    /**
+     * All physical parameters for physical SiDB simulations.
+     */
+    const sidb_simulation_parameters& sidb_sim_params{};
+    /**
+     * Probability that the ground state is less populated due to temperature. For gate-based simulation, this is the
+     * probability of erroneous calculations of the gate.
+     */
+    double confidence_level{0.99};
+    /**
+     * Simulation stops at max_temperature (room temperature ~300 K).
+     */
+    uint64_t max_temperature{400};
+    /**
+     * Truth table of the given gate (if layout is simulated in `gate-based` mode).
+     */
+    tt truth_table{};
+    /**
+     * Input bit (e.g. 0 -> 00, 1 -> 01, ...).
+     */
+    uint64_t input_bit{};
+};
+
+/**
+ * This struct stores the result of the temperature simulation.
+ *
+ * @paramt Lyt Cell-level layout type.
+ */
+template <typename Lyt>
+struct critical_temperature_stats
```

Comment:
Are we not counting the runtime of this algorithm?

Reply:
I think it is not important for this algorithm since the runtime is mainly dominated by the exhaustive simulation algorithm used. But please disagree if you think we need the runtime.

Reply:
If that is the case, we indeed do not need runtime measurement